docs/design/proposals/collections.md

```diff
@@ -0,0 +1,633 @@
+Kubernetes resource. If you need 5 Pods, you write 5 resource blocks. If you
+later need 10, you must modify the RGD to add 5 more blocks.
+
+Consider a batch processing application where users specify how many worker Pods
+to run through their custom resource: `workers: 5`. Each worker needs its own
+Pod with specific environment variables and configurations based on its index.
+With KRO today, you cannot express "create N Pods where N equals spec.workers".
+The same limitation appears in data processing pipelines where each stage might
+need a different number of resources:
+`stages: [{name: "extract", replicas: 3}, {name: "transform", replicas: 5}, {name: "load", replicas: 2}]`.
+
+This forces users to either hardcode a fixed number of resources and live with
+the inflexibility, or wrap KRO with templating tools to generate different
+RGDs - defeating KRO's purpose of moving beyond templating to declarative APIs.
+
+What users really need is the ability to say "for each item in this array,
+create a resource" or "create N resources where N comes from the spec" - dynamic
+resource creation based on runtime values while maintaining KRO's declarative
+approach.
+
+## Proposal
+
+### Overview
+
+This KREP introduces the `forEach` field for resource definitions, enabling a
+single resource block to create multiple resources based on collections. The
+forEach field can iterate over:
+
+- Arrays from the custom resource spec
+- Ranges of numbers, characters, or other types
+- Collections of discovered resources (via externalRef)
+- Outputs from other resources in the RGD
+
+Each iteration creates a separate node in KRO's DAG, maintaining clear ownership
+and reconciliation semantics. The number of DAG nodes is determined at runtime
+but execution remains bounded and predictable.
+
+### Core Concepts
+
+**Collections as Declarative Relationships**
+
+The `forEach` field establishes a declarative relationship between a collection
+and a set of resources. When you write `forEach: ${spec.workers}`, you're
+declaring "there should exist one resource for each item in spec.workers". KRO
+maintains this relationship through continuous reconciliation, creating and
+deleting resources as the collection changes.
+
+**Bounded Execution**
+
+While `forEach` enables dynamic resource creation, it maintains KRO's security
+guarantees. The size of collections is bounded by Kubernetes API limits, CEL
+expressions within `forEach` are subject to normal CEL cost limits, and the
+total number of resources created is visible and predictable before execution
+begins.
+
+**DAG Expansion**
+
+A resource block with the `forEach` field expands into multiple nodes in KRO's
+execution DAG at runtime. Each expanded node is treated as a distinct resource
+with its own reconciliation cycle, status tracking, and error handling.
+Dependencies between resources are preserved - if resource B depends on resource
```

Comment:
Would there be a way to depend on a subset of them? Or it's always all or nothing?